### PR TITLE
feat: test workflow cron job

### PR DESCRIPTION
**Pull request by vsukhin**
## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [ ] tested locally
- [x] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

## Breaking changes

-

## Changes

-

## Fixes

-

### CODE REVIEW

File: api/testworkflows/v1/types.go [hunk 1]
```diff
@@ -72,3 +72,17 @@ type PodConfig struct {
...
 }
+
+type Event struct {
+	Cronjob *CronJobConfig `json:"cronjob,omitempty"`
+}
+
+// cron job configuration
+type CronJobConfig struct {
+	// cron schedule to run a test workflow
+	Cron string `json:"cron" expr:"template"`
```
Comment (rangoo94) [inline api/testworkflows/v1/types.go:83]: Probably worth to allow passing optional `config` as well, that would be used to configure such TestWorkflow under this condition

File: internal/controller/testworkflows/testworkflow_controller.go [hunk 2]
```diff
@@ -46,9 +53,162 @@ type TestWorkflowReconciler struct {
...
+				Group:       testworkflowsv1.Group,
+				Resource:    testworkflowsv1.Resource,
+				Version:     testworkflowsv1.Version,
+				ResourceURI: cronjob.TestWorkflowResourceURI,
+				Labels:      newCronJobConfig.Labels,
+				Annotations: newCronJobConfig.Annotations,
+				Data:        "[]",
+			}
+
+			if err = r.CronJobClient.Create(ctx, testWorkflow.Name,
```
Comment (rangoo94) [inline internal/controller/testworkflows/testworkflow_controller.go:181]: The name is unique, while it should rather allow multiple cronjobs for a single TestWorkflow - probably worth to add event index to name

File: internal/controller/testworkflows/testworkflow_controller.go [hunk 3]
```diff
@@ -46,9 +53,162 @@ type TestWorkflowReconciler struct {
...
+	if !ok && hasTemplates {
+		if testWorkflow.Labels == nil {
+			testWorkflow.Labels = make(map[string]string)
+		}
+
+		testWorkflow.Labels[cronjob.TestWorkflowTemplateResourceURI] = "yes"
+		return ctrl.Result{}, r.Update(ctx, &testWorkflow)
+	}
+
+	newCronJobConfigs := make(map[string]*testworkflowsv1.CronJobConfig)
```
Comment (rangoo94) [inline internal/controller/testworkflows/testworkflow_controller.go:103]: Not sure, but I think that it should be rather a list - you may want to have multiple cronjobs with the same schedule, but different config parameters

File: api/testworkflows/v1/types.go (hunk 1 above)
Comment (vsukhin) [reply api/testworkflows/v1/types.go:83]: From what I got from discussion we can add it in next iteration

File: internal/controller/testworkflows/testworkflow_controller.go (hunk 3 above)
Comment (vsukhin) [reply internal/controller/testworkflows/testworkflow_controller.go:103]: If we add config parameters, then we will need to switch to list

Comment (vsukhin) [reply internal/controller/testworkflows/testworkflow_controller.go:103]: yes, like, we will delete the old cron job without config and create a few new ones with configs, or might be, keep old one if we have cron jobs with emoty configs and add new ones with configs

**Review by rangoo94 — APPROVED** [review]

**Pull request merged**